supervision/detection/utils.py

```diff
@@ -110,6 +110,196 @@ def non_max_suppression(
+
+    return keep_to_merge_list
+
+
+def batched_greedy_nmm(
+    predictions: np.ndarray, threshold: float = 0.5
+) -> Dict[int, List[int]]:
+    """
+    Apply greedy version of non-maximum merging per category to avoid detecting
+    too many overlapping bounding boxes for a given object.
+
+    Args:
+        predictions (np.ndarray): An array of shape `(n, 6)` containing
+            the bounding boxes coordinates in format `[x1, y1, x2, y2]`,
+            the confidence scores and class_ids.
+        threshold (float, optional): The intersection-over-union threshold
+            to use for non-maximum suppression. Defaults to 0.5.
+
+    Returns:
+        Dict[int, List[int]]: Mapping from prediction indices
+        to keep to a list of prediction indices to be merged.
+    """
+    category_ids = predictions[:, 5]
+    keep_to_merge_list = {}
+    for category_id in np.unique(category_ids):
+        curr_indices = np.where(category_ids == category_id)[0]
+        curr_keep_to_merge_list = greedy_nmm(predictions[curr_indices], threshold)
+        curr_indices_list = curr_indices.tolist()
+        for curr_keep, curr_merge_list in curr_keep_to_merge_list.items():
+            keep = curr_indices_list[curr_keep]
+            merge_list = [curr_indices_list[i] for i in curr_merge_list]
+            keep_to_merge_list[keep] = merge_list
+    return keep_to_merge_list
+
+
+def get_merged_bbox(bbox1: np.ndarray, bbox2: np.ndarray) -> np.ndarray:
+    """
+    Merges two bounding boxes into one.
+
+    Args:
+        bbox1 (np.ndarray): A numpy array of shape `(, 4)` where the
+            row corresponds to a bounding box in
+            the format `(x_min, y_min, x_max, y_max)`.
+        bbox2 (np.ndarray): A numpy array of shape `(, 4)` where the
+            row corresponds to a bounding box in
+            the format `(x_min, y_min, x_max, y_max)`.
+
+    Returns:
+        np.ndarray: A numpy array of shape `(, 4)` where the new
+            bounding box is the merged bounding box of `bbox1` and `bbox2`.
+    """
+    left_top = np.minimum(bbox1[0][:2], bbox2[0][:2])
+    right_bottom = np.maximum(bbox1[0][2:], bbox2[0][2:])
+    return np.array([np.concatenate([left_top, right_bottom])])
+
+
+def get_merged_class_id(id1: np.ndarray, id2: np.ndarray) -> np.ndarray:
+    """
+    Merges two class ids into one.
+
+    Args:
+        id1 (np.ndarray): The first class id.
+        id2 (np.ndarray): The second class id.
+
+    Returns:
+        np.ndarray: The merged class id.
+    """
+    return np.array([max(id1.item(), id2.item())])
+
+
+def get_merged_confidence(
+    confidence1: np.ndarray, confidence2: np.ndarray
+) -> np.ndarray:
+    """
+    Merges two confidences into one.
+
+    Args:
+        confidence1 (np.ndarray): The first confidence.
+        confidence2 (np.ndarray): The second confidence.
+
+    Returns:
+        np.ndarray: The merged confidence.
+    """
+    return np.array([max(confidence1.item(), confidence2.item())])
```

Comment:
We aren't as confident about the larger area - there's outlier cases where this is dangerous.

In theory, is there a way to draw weights from this - to know which area contributed how much; to derive the weights?
Alternatively, I think `mean` of the two is more acceptable than `max`.

Reply:
I don't quite get this. Where does your 'area' come from?
Should the merged confidence still be `mean` or `max`, since we are using the more confident Detection to determine the `data`, `class_id` and `tracker_id`?

Reply:
Just to clarify - NMM doesn't give a number to say how much each area contributed to the new merged object, right? If it did, we could multiply that with confidences and get a normalized new confidence.

If it doesn't have that, I'll make a judgement call - let's use `mean`. Yes, this is inconsistent with how we get `data`and `ids`, but it's better than `max` which is equivalent to saying that we're just as confident about the newly merged area as we were about any of its parts.

Reply:
Got it!
We do have the bounding boxes for each detection pair that is getting merged.
Using the merged bounding box, I could get the relative area each individual bounding box occupies in the merged bounding box to calculate the contribution of each object.
Does that sound like a solution or should we just stick to `mean`?